### PR TITLE
New module D.TwoD.Layout.Constrained

### PR DESCRIPTION
**Pull request by byorgey**
Generated Haddocks are here: http://ozark.hendrix.edu/~yorgey/hosted/diagrams-contrib/Diagrams-TwoD-Layout-Constrained.html

### CODE REVIEW

File: src/Diagrams/TwoD/Layout/Constrained.hs [hunk 1]
```diff
@@ -0,0 +1,671 @@
...
+------------------------------------------------------------
+-- Variables and expressions
+------------------------------------------------------------
+
+-- | An abstract type representing unique IDs for diagrams.  The
+--   constructor is not exported, so the only way to obtain a 'DiaID'
+--   is by calling 'newDia' or 'newDias'. The phantom type parameter
+--   's' ensures that such 'DiaID's can only be used with the
+--   constrained system in which they were introduced.
+newtype DiaID s = DiaID Int
```
Comment (cchalmers) [inline src/Diagrams/TwoD/Layout/Constrained.hs]: This (as well as `ConstrainedState`) isn't exported, so you can't see the documentation in haddock. Of course you don't want to export the constructors but I think the documentation is helpful.

File: src/Diagrams/TwoD/Layout/Constrained.hs [hunk 2]
```diff
@@ -0,0 +1,671 @@
...
+-- | Introduce a new diagram into the constrained system.  Returns a
+--   unique ID for use in referring to the diagram later.
+--
+--   Both the position of the diagram's origin and its scaling factor
+--   may be constrained.  If unconstrained, the origin will default to
+--   (0,0), and the scaling factor to 1, respectively.
+newScalableDia :: QDiagram b V2 n m -> Constrained s b n m (DiaID s)
+newScalableDia dia = do
+  d <- DiaID <$> (diaCounter <+= 1)
+  diagrams %= (\m -> m & L.at d .~ Just dia)
```
Comment (cchalmers) [inline src/Diagrams/TwoD/Layout/Constrained.hs]: `diagrams . L.at d ?= dia`?

**Comment by cchalmers** [comment]
I haven't had a chance to try it out yet but it looks pretty cool. I'll give it a go when I get a chance.

File: src/Diagrams/TwoD/Layout/Constrained.hs (hunk 1 above)
Comment (byorgey) [inline src/Diagrams/TwoD/Layout/Constrained.hs]: Good point.  I'll export them.

File: src/Diagrams/TwoD/Layout/Constrained.hs (hunk 2 above)
Comment (byorgey) [inline src/Diagrams/TwoD/Layout/Constrained.hs]: Oh, thanks, I didn't know about `?=`.

**Comment by bergey** [comment]
I'm looking forward to using this!

The [build failures for GHC-7.10](https://travis-ci.org/diagrams/diagrams-contrib/jobs/101148329) look like you meant to export the generated lenses for `Var`.  (Failures for 7.8 are the same.)

The HP-2014 build is failing because HP-2014 has `mtl-2.1.3.1` and `mfsolve` requires `mtl>=2.2.1`  Do you want to drop support for HP-2014, or get @kuribas to allow older `mtl`?

**Comment by kuribas** [comment]
@bergey done

**Comment by bergey** [comment]
@kuribas Thanks!

**Comment by byorgey** [comment]
Hah, great, thanks @kuribas !

Hopefully I have now taken care of the build failures.  I don't actually want to export the lenses (`Var` should be abstract), so I just changed it to only generate the one lens I actually do use.

**Comment by byorgey** [comment]
@kuribas : see https://github.com/kuribas/mfsolve/issues/6

**Pull request merged**